cloudstate/discovery/server.go

```diff
@@ -0,0 +1,184 @@
+func (s *EntityDiscoveryServer) ReportError(_ context.Context, error *protocol.UserFunctionError) (*empty.Empty, error) {
+	log.Printf("ReportError: %v\n", error)
+	return &empty.Empty{}, nil
+}
+
+func (s *EntityDiscoveryServer) updateSpec() (err error) {
+	protoBytes, err := proto.Marshal(s.fileDescriptorSet)
+	if err != nil {
+		return errors.New("unable to Marshal FileDescriptorSet")
+	}
+	s.entitySpec.Proto = protoBytes
+	return nil
+}
+
+func (s *EntityDiscoveryServer) resolveFileDescriptors(config protocol.DescriptorConfig) error {
+	if config.Service != "" {
+		if err := s.registerFileDescriptorProto(config.Service); err != nil {
+			return err
+		}
+	}
+	// Add dependent domain descriptors.
+	for _, dp := range config.Domain {
+		if err := s.registerFileDescriptorProto(dp); err != nil {
+			return err
+		}
+	}
+	for _, dm := range config.DomainMessages {
+		if err := s.registerFileDescriptor(dm); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (s *EntityDiscoveryServer) RegisterEventSourcedEntity(entity *eventsourced.Entity, config protocol.DescriptorConfig) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if err := s.resolveFileDescriptors(config); err != nil {
+		return fmt.Errorf("failed to resolve FileDescriptor for DescriptorConfig: %+v: %w", config, err)
+	}
+	s.entitySpec.Entities = append(s.entitySpec.Entities, &protocol.Entity{
+		EntityType:    protocol.EventSourced,
+		ServiceName:   entity.ServiceName.String(),
+		PersistenceId: entity.PersistenceID,
+	})
+	return s.updateSpec()
+}
+
+func (s *EntityDiscoveryServer) RegisterCRDTEntity(entity *crdt.Entity, config protocol.DescriptorConfig) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if err := s.resolveFileDescriptors(config); err != nil {
+		return fmt.Errorf("failed to resolveFileDescriptor for DescriptorConfig: %+v: %w", config, err)
+	}
+	s.entitySpec.Entities = append(s.entitySpec.Entities, &protocol.Entity{
+		EntityType:  protocol.CRDT,
+		ServiceName: entity.ServiceName.String(),
+	})
+	return s.updateSpec()
+}
+
+func (s *EntityDiscoveryServer) hasRegistered(filename string) bool {
+	for _, f := range s.fileDescriptorSet.File {
+		if f.GetName() == filename {
+			return true
+		}
+	}
+	return false
+}
+
+func (s *EntityDiscoveryServer) registerFileDescriptorProto(filename string) error {
+	if s.hasRegistered(filename) {
+		return nil
+	}
+	desc, err := unpackFile(proto.FileDescriptor(filename))
+	if err != nil {
+		return fmt.Errorf("failed to registerFileDescriptorProto for filename: %s: %w", filename, err)
+	}
+	s.fileDescriptorSet.File = append(s.fileDescriptorSet.File, desc)
+	for _, dep := range desc.Dependency {
+		err := s.registerFileDescriptorProto(dep)
+		if err != nil {
+			return err
+		}
+	}
+	return s.updateSpec()
+}
+
+func (s *EntityDiscoveryServer) registerFileDescriptor(msg descriptor.Message) error {
+	fd, _ := descriptor.ForMessage(msg) // this can panic.
+	if r := recover(); r != nil {
```

Comment:
recover has to be in a defer block

Reply:
@chongkai great catch!
We even can let it panic I think, as `descriptor.ForMessage` does not document that it crashes at all.

The deprecated use of `descriptor.ForMessage` will be mitigated with the switch over to protobuf API v2 with https://github.com/cloudstateio/go-support/issues/20